src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -352,64 +396,506 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+	  }
+	  
+	  cb->handle_data(bl, ofs, len);
+	}
+      } else {
+	/ Backend store retrieval /
+	ldpp_dout(dpp, 20) << "D4N Filter: Object directory get operation failed." << dendl;
+	ret = next->iterate(dpp, ofs, end, cb, y);
+
+	if (!ret) {
+	  / Cache block locally /
+	  ret = source->driver->get_cache_driver()->put(dpp, source->get_key().get_oid(), bl, len, source->get_attrs()); // May be put_async -Sam
+
+	  / Store block in directory /
+	  rgw::d4n::BlockDirectory* tempBlockDir = source->driver->get_block_dir(); // remove later -Sam
+
+	  source->driver->get_cache_block()->hostsList.push_back(tempBlockDir->get_addr().host + ":" + std::to_string(tempBlockDir->get_addr().port)); // local cache address -Sam 
+	  source->driver->get_cache_block()->size = source->get_obj_size();
+	  source->driver->get_cache_block()->cacheObj.bucketName = source->get_bucket()->get_name();
+	  source->driver->get_cache_block()->cacheObj.objName = source->get_key().get_oid();
+
+	  int setDirReturn = tempBlockDir->set_value(source->driver->get_cache_block());
+
+	  if (setDirReturn < 0) {
+	    ldpp_dout(dpp, 20) << "D4N Filter: Block directory set operation failed." << dendl;
+	  } else {
+	    ldpp_dout(dpp, 20) << "D4N Filter: Block directory set operation succeeded." << dendl;
+	  }
+	}
+      }
+    }
+  }
+
+  if (ret < 0) 
+    ldpp_dout(dpp, 20) << "D4N Filter: Cache iterate operation failed." << dendl;
+
+  return next->iterate(dpp, ofs, end, cb, y); */
+}
+
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::flush_last_part(const DoutPrefixProvider* dpp)
+{
+  save_dpp = dpp;
+  last_part = true;
+  return handle_data(bl_rem, 0, bl_rem.length());
+}
+
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  auto rgw_get_obj_max_req_size = g_conf()->rgw_get_obj_max_req_size;
+
+  if (!last_part && bl.length() <= rgw_get_obj_max_req_size) {
+    auto r = client_cb->handle_data(bl, bl_ofs, bl_len);
+
+    if (r < 0) {
+      return r;
+    }
+  }
+
+  int ret = 0;
+  //Accumulating data from backend store into rgw_get_obj_max_req_size sized chunks and then writing to cache
+  if (write_to_cache) {
+    const std::lock_guard l(d3n_get_data.d3n_lock);
+    rgw::d4n::CacheBlock block;
+    rgw::d4n::BlockDirectory* blockDir = source->driver->get_block_dir();
+    if (bl.length() > 0 && last_part) { // if bl = bl_rem has data and this is the last part, write it to cache
+      std::string oid = this->oid + "_" + std::to_string(ofs) + "_" + std::to_string(bl_len);
+      ret = filter->get_cache_driver()->put_async(save_dpp, oid, bl, bl.length(), source->get_attrs());
+      if (ret == 0) {
+        block.size = bl.length();
+        block.blockId = oid;
+        block.hostsList.push_back(blockDir->get_addr().host + ":" + std::to_string(blockDir->get_addr().port)); //check if this is the correct way to get the local addr
+        block.size = source->get_obj_size();
```

Comment:
Redeclaration of `block.size`

Reply:
Yes - block.size must *NOT* be source->get_obj_size();, please remove this in your PR.